src/librustc_target/abi/mod.rs

```diff
@@ -1081,4 +1119,82 @@ impl<'a, Ty> TyLayout<'a, Ty> {
             Abi::Aggregate { sized } => sized && self.size.bytes() == 0
         }
     }
+
+    /// Determines if this type permits "raw" initialization by just transmuting some
+    /// memory into an instance of `T`.
+    /// `zero` indicates if the memory is zero-initialized, or alternatively
+    /// left entirely uninitialized.
+    /// This is conservative: in doubt, it will answer `true`.
+    pub fn might_permit_raw_init<C, E>(
+        self,
+        cx: &C,
+        zero: bool,
+    ) -> Result<bool, E>
+    where
+        Self: Copy,
+        Ty: TyLayoutMethods<'a, C>,
+        C: LayoutOf<Ty = Ty, TyLayout: MaybeResult<Self, Error = E>>
+    {
+        let scalar_allows_raw_init = move |s: &Scalar| -> bool {
+            let range = &s.valid_range;
+            if zero {
+                // The range must contain 0.
+                range.contains(&0) ||
+                (*range.start() > *range.end()) // wrap-around allows 0
+            } else {
+                // The range must include all values.
+                *range.start() == range.end().wrapping_add(1)
```

Comment:
You should probably call `valid_range_exclusive` and check that the resulting range has the same `start` and `end` (to avoid having to write masking code yourself here).

Reply:
*Oh*, that is a useful method. Any reason we shouldn't use it for the zero test as well, and for Miri's validity checking?

Reply:
The returned range is still a wrap-around one, so it doesn't help you unless you 1. need an exclusive wrap-around range to give to LLVM, or you 2. need to check if the range covers all the bitpatterns of that size.

Reply:
Oh wait, this still has wrap-around. Never mind.